pclntab/pclntab.go

```diff
@@ -0,0 +1,652 @@
+		funcNpcdataOffset:   funcNpcdataOffset,
+		funcNfuncdataOffset: funcNfuncdataOffset,
+		functab:             functab,
+		funcdata:            funcdata,
+	}, nil
+}
+
+func FindGoPCLnTab(ef *pfelf.File) (goPCLnTabInfo *GoPCLnTabInfo, err error) {
+	if !disableRecover {
+		// gopclntab parsing code might panic if the data is corrupt.
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("panic while searching pclntab: %v", r)
+			}
+		}()
+	}
+
+	var data []byte
+	var goPCLnTabAddr uint64
+	var symtab *libpf.SymbolMap
+
+	goPCLnTabEndKnown := true
+	if s := ef.Section(".gopclntab"); s != nil {
+		if data, err = s.Data(maxBytesGoPclntab); err != nil {
+			return nil, fmt.Errorf("failed to load .gopclntab: %w", err)
+		}
+		goPCLnTabAddr = s.Addr
+	} else if s := ef.Section(".data.rel.ro.gopclntab"); s != nil {
+		if data, err = s.Data(maxBytesGoPclntab); err != nil {
+			return nil, fmt.Errorf("failed to load .data.rel.ro.gopclntab: %w", err)
+		}
+		goPCLnTabAddr = s.Addr
+	} else if s := ef.Section(".go.buildinfo"); s != nil {
+		symtab, err = ef.ReadSymbols()
+		if err != nil {
+			// It seems the Go binary was stripped, use the heuristic approach to get find gopclntab.
+			// Note that `SearchGoPclntab` returns a slice starting from gopcltab header to the end of segment
+			// containing gopclntab. Therefore this slice might contain additional data after gopclntab.
+			// There does not seem to be an easy way to get the end of gopclntab segment without parsing the
+			// gopclntab itself.
+			if data, goPCLnTabAddr, err = SearchGoPclntab(ef); err != nil {
+				return nil, fmt.Errorf("failed to search .gopclntab: %w", err)
+			}
+			// Truncate the data to the end of the section containing gopclntab.
+			if sec := sectionContaining(ef, goPCLnTabAddr); sec != nil {
+				data = data[:sec.Addr+sec.Size-goPCLnTabAddr]
+			}
+			goPCLnTabEndKnown = false
+		} else {
+			var start, end libpf.SymbolValue
+			start, err = symtab.LookupSymbolAddress("runtime.pclntab")
+			if err != nil {
+				return nil, fmt.Errorf("failed to load .gopclntab via symbols: %w", err)
+			}
+			end, err = symtab.LookupSymbolAddress("runtime.epclntab")
+			if err != nil {
+				return nil, fmt.Errorf("failed to load .gopclntab via symbols: %w", err)
+			}
+			if start >= end {
+				return nil, fmt.Errorf("invalid .gopclntab symbols: %v-%v", start, end)
+			}
+			data = make([]byte, end-start)
+			if _, err = ef.ReadVirtualMemory(data, int64(start)); err != nil {
+				return nil, fmt.Errorf("failed to load .gopclntab via symbols: %w", err)
+			}
+			goPCLnTabAddr = uint64(start)
+		}
+	}
+
+	if data == nil {
+		return nil, nil
+	}
+
+	goPCLnTabInfo, err = parseGoPCLnTab(data)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse .gopclntab: %w", err)
+	}
+
+	goPCLnTabInfo.Address = goPCLnTabAddr
+
+	if !goPCLnTabEndKnown && goPCLnTabInfo.Version >= ver116 {
+		funcDataSize := goPCLnTabInfo.findFuncDataSize()
+		if funcDataSize != -1 {
```

Comment:
Could we add a comment here to indicate why it's okay not to truncate the data in that case?

Reply:
Done